### PR TITLE
Optimization run

### PR DESCRIPTION
**Pull request by sirenkovladd**
remove condition from run
check isAsync with `fn() instanceof Promise;`

### CODE REVIEW

File: src/task.ts [hunk 1]
```diff
@@ -74,13 +88,7 @@ export default class Task extends EventTarget {
...
         }
 
-        const taskStart = this.bench.now();
-        if (isAsync) {
-          await this.fn();
-        } else {
-          this.fn();
-        }
-        const taskTime = this.bench.now() - taskStart;
+        const taskTime = isAsync ? await this.runAsync() : this.runSync();
```
Comment (kibertoad) [inline src/task.ts]: why this change? doesn't it do exactly same thing but with more code?

File: src/utils.ts [hunk 2]
```diff
@@ -25,4 +25,18 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
 /**
  * an async function check method only consider runtime support async syntax
  */
-export const isAsyncFunction = (fn: Fn) => fn.constructor === AsyncFunctionConstructor;
+export const isAsyncFunction = (fn: Fn) => {
+  if (fn.constructor === AsyncFunctionConstructor) {
+    return true;
+  }
+  try {
+    const call = fn();
```
Comment (kibertoad) [inline src/utils.ts]: wouldn't that produce side effects?

Comment (sirenkovladd) [reply src/utils.ts]: Well, the function is designed to be called a large number of times, right?

File: src/task.ts (hunk 1 above)
Comment (sirenkovladd) [reply src/task.ts]: previously, the condition itself was included in the challenge
for quick functions it makes sense

Comment (kibertoad) [reply src/task.ts]: What do you mean? Isn't exactly the same code going to be executed the same amount of times now, just spread differently?

Comment (sirenkovladd) [reply src/task.ts]: In my changes, the calculation is in runAsync/runSync
I mean the execution of the condition is above the time calculation, not in the middle

Comment (kibertoad) [reply src/task.ts]: @sirenkovladd I don't see that. Previously it calculated starting point, executed function, calculated taskTime. It does exactly same thing now.

unless you refer to `isAsync` change, which did not require the rest of the changes.

Comment (sirenkovladd) [reply src/task.ts]: https://github.com/tinylibs/tinybench/blob/4d2c04f564586bdb3c8914fee54117a8faa0fc9e/src/task.ts#L77-L83

I see a condition in the calculation of time

File: src/utils.ts (hunk 2 above)
Comment (sirenkovladd) [reply src/utils.ts]: Do you have one, do you have an example?

File: src/task.ts (hunk 1 above)
Comment (sirenkovladd) [reply src/task.ts]: If you want I can take out the `while` from the `run` in `runAsync`/`run/Sync`

**Comment by Aslemammad** [comment]
I like this change, I see it removing overhead, but did you see that in the results? Have you been able to notice any thing? I'd like to see it, I'm just curious. 
This PR LGTM, but I'd like to see @mcollina and @Uzlopak's opinion on this.

File: src/utils.ts [hunk 3]
```diff
@@ -25,4 +25,24 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
 /**
  * an async function check method only consider runtime support async syntax
  */
-export const isAsyncFunction = (fn: Fn) => fn.constructor === AsyncFunctionConstructor;
+export const isAsyncFunction = async (task: Task) => {
+  if (task.fn.constructor === AsyncFunctionConstructor) {
```
Comment (Aslemammad) [inline src/utils.ts]: isn't this condition only enough?

Comment (Aslemammad) [reply src/utils.ts]: I guess with the new condition it determines functions that return promises? right?

Comment (sirenkovladd) [reply src/utils.ts]: Not detectable

```js
.add('foo', () => {
      return new Promise((resolve) => setTimeout(resolve, 100));
    })
```

Comment (Aslemammad) [reply src/utils.ts]: Ok, I get it now. It's reasonable.

**Comment by Uzlopak** [comment]
isAsyncFunction was doing totally perfect sense, as it was fast to determine if we have a function of type `async function () {}`. 

Now isAsyncFunction makes one run of the function and determines various types of async behaviour. Maybe make it two function. Restore the previous isAsyncFunction and create a new isAsyncTask, which does all the magic.

Also, the `call instanceof Promise` is not enough. Yes it detects built-in Promises. But It should handle PromiseLike. So `call && typeof call.then === 'function'`

**Comment by Uzlopak** [comment]
@Aslemammad 

Please check discord.

**Comment by sirenkovladd** [comment]
> But It should handle PromiseLike So `call && typeof call.then === 'function'`

I agree with you, I will change it

File: src/utils.ts [hunk 4]
```diff
@@ -25,17 +26,19 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
...
   }
   try {
     if (task.opts.beforeEach != null) {
       await task.opts.beforeEach.call(task);
     }
     const call = task.fn();
-    const result = call instanceof Promise;
-    if (result) {
+    const result = typeof call?.then === 'function';
+    if (result && call?.catch) {
```
Comment (Uzlopak) [inline src/utils.ts]: You could actually send it to nirvana by doing

call.then(() => {}, () => {})

So you can save your check for catch. E.g. some fastify classes are thenable but have no .catch.

File: src/utils.ts [hunk 5]
```diff
@@ -25,17 +26,19 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
...
     return true;
   }
   try {
     if (task.opts.beforeEach != null) {
       await task.opts.beforeEach.call(task);
     }
     const call = task.fn();
-    const result = call instanceof Promise;
-    if (result) {
+    const result = typeof call?.then === 'function';
```
Comment (Uzlopak) [inline src/utils.ts]: Maybe `result` is too generic. Actually should be renamed to `promiseLike`.

Maybe create a typeguard isPromiseLike

See 
https://github.com/fastify/avvio/blob/master/lib/is-promise-like.js

File: src/utils.ts (hunk 4 above)
Comment (sirenkovladd) [reply src/utils.ts]: > call.then(() => {}, () => {})

it is just `await call`

You can give some example or give some test that will reject

Comment (Uzlopak) [reply src/utils.ts]: Then why do you check for catch?

File: src/utils.ts (hunk 5 above)
Comment (sirenkovladd) [reply src/utils.ts]: done it

File: src/utils.ts (hunk 4 above)
Comment (sirenkovladd) [reply src/utils.ts]: Some tests throw unhandled errors

Comment (sirenkovladd) [reply src/utils.ts]:
```js
.add('error', () => {
      return Promise.reject(new Error('fake'));
    })
```

File: src/utils.ts [hunk 6]
```diff
@@ -26,3 +35,29 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
...
  */
 export const isAsyncFunction = (fn: Fn) => fn.constructor === AsyncFunctionConstructor;
+
+export const isAsyncTask = async (task: Task) => {
+  if (isAsyncFunction(task.fn)) {
+    return true;
+  }
+  try {
+    if (task.opts.beforeEach != null) {
+      await task.opts.beforeEach.call(task);
```
Comment (Uzlopak) [inline src/utils.ts]: hmm

kind of feels error prone. if beforeEach or afterEach fail, then we result in the fact that the Task is not async. 

imho beforeEach and afterEach have to be in try catch blocks

File: test/index.test.ts [hunk 7]
```diff
@@ -316,3 +316,14 @@ test('task beforeAll, afterAll, beforeEach, afterEach', async () => {
   expect(afterEach.mock.calls.length).toBe(iterations * 2);
   expect(beforeEach.mock.calls.length).toBe(afterEach.mock.calls.length);
 });
+
+test('task with promiseLike return', async () => {
```
Comment (Uzlopak) [inline test/index.test.ts:320]: we are lacking a test, where a sync function returns a promise

File: src/utils.ts (hunk 6 above)
Comment (sirenkovladd) [reply src/utils.ts]: good point, thank you

File: test/index.test.ts (hunk 7 above)
Comment (sirenkovladd) [reply test/index.test.ts:320]: This test actually does that

Comment (Uzlopak) [reply test/index.test.ts:320]: Why do we then bother to cover various types if we dont test for them.

This is what I want as an additional test (not tested!):

```js
test('sync task which returns a promise', async () => {
  const bench = new Bench();

  bench.add('foo', () => {
    return new Promise(resolve => setTimeout(resolve, 100))
  });

  await bench.run();
  expect(bench.getTask('foo')!.result!.mean).toBeGreaterThan(100);
});

Comment (sirenkovladd) [reply test/index.test.ts:320]: Okay, I'll add, it's the same for me

```js
bench.add('foo', () => ({
    then: async (resolve: () => void) => setTimeout(resolve, 100),
  }));
```

Comment (Uzlopak) [reply test/index.test.ts:320]: still, add that case please.

**Comment by Uzlopak** [comment]
consider creating a new test file, where you test all aspects of isAsyncTask.

Especially write unit tests, that check if isAsyncTask returns correct values if beforeEach and/or afterEach throw an error.

**Comment by sirenkovladd** [comment]
@Uzlopak, that's all?

**Review by Uzlopak — APPROVED** [review]
LGTM

But waiting for feedback by @kibertoad and @Aslemammad

**Review by Aslemammad — APPROVED** [review]

**Comment by sirenkovladd** [comment]
Do you have any blocks to merge PR?

**Comment by Uzlopak** [comment]
I would prefer that @kibertoad have a look. I now would just say, that he has time till sunday. If on sunday no feedback is given, we should not block it longer.

File: src/task.ts (hunk 1 above)
Comment (kibertoad) [reply src/task.ts]: I feel that we are talking about very different things. Can we hop to discord to discuss in realtime, perhaps?

File: src/task.ts [hunk 8]
```diff
@@ -95,7 +107,11 @@ export default class Task extends EventTarget {
     }
 
     if (this.opts.afterAll != null) {
-      await this.opts.afterAll.call(this);
+      try {
+        await this.opts.afterAll.call(this);
+      } catch (e) {
```
Comment (kibertoad) [inline src/task.ts:107]: is this change related to optimizations?

File: src/utils.ts [hunk 9]
```diff
@@ -26,3 +35,37 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
...
+
+export const isAsyncTask = async (task: Task) => {
+  if (isAsyncFunction(task.fn)) {
+    return true;
+  }
+  try {
+    if (task.opts.beforeEach != null) {
+      try {
+        await task.opts.beforeEach.call(task);
+      } catch (e) {
```
Comment (kibertoad) [inline src/utils.ts:47]: why do we ignore thrown errors here?

Comment (sirenkovladd) [reply src/utils.ts:47]: Because this function is only responsible for determining whether this function is synchronous

File: src/task.ts (hunk 8 above)
Comment (sirenkovladd) [reply src/task.ts:107]: no, but because of this there will be an error in the tests

**Comment by sirenkovladd** [comment]
Do you have any other questions about PR?

**Comment by kibertoad** [comment]
@sirenkovladd Everything else looks good! Can we remove the unnecessary code duplication, though? I still don't see what it adds.

**Comment by sirenkovladd** [comment]
Sorry, what do you mean by `unnecessary code duplication`?

**Comment by kibertoad** [comment]
@sirenkovladd 

```ts
  private runSync() {
    const taskStart = this.bench.now();
    this.fn();
    const taskTime = this.bench.now() - taskStart;
    return taskTime;
  }

  private async runAsync() {
    const taskStart = this.bench.now();
    await this.fn();
    const taskTime = this.bench.now() - taskStart;
    return taskTime;
  }
```

This is exactly same code, the only difference is whether we await it or not. This can be handled as a simple condition, without copy-pasting the rest of it.

**Comment by sirenkovladd** [comment]
OK, i got it
but the point is that there are no conditions when calculating the speed of execution (if i add `isSync ? call() : await call()`)
also if I call `await` when calling synchronous code (to combine these two cases) it will impose extra time on the synchronous function (already tested) which will make the calculation not correct

so I think this is the best way to calculate the execution time correctly

**Comment by kibertoad** [comment]
OK, I see your point. Not sure that's the tradeoff that I would make, but I see your reasoning, and it's sound.

**Review by kibertoad — APPROVED** [review]

**Comment by Uzlopak** [comment]
wait

**Comment by Uzlopak** [comment]
just for clarification: 

we should not have a condition inside the code, where we started the benchmark. 

So something you proposed 

 ```js
const taskStart = this.bench.now(); 
if (isAsync) { 
	 await this.fn(); 
} else { 
	 this.fn(); 
}
const taskTime = this.bench.now() - taskStart; 
```

would be imho bad. Thats why I am ok with the current solution. 

What we actually just need to do is following. 

 ```js
let taskTime = 0
if (isAsync) {
    const taskStart = this.bench.now();
    await this.fn();
    taskTime = this.bench.now() - taskStart; 
} else {
    const taskStart = this.bench.now();
    this.fn();
    taskTime = this.bench.now() - taskStart; 
}
```

then we dont need the private methods runAsync and runSync, which btw. imho are not a useful abstraction.

**Comment by Uzlopak** [comment]
Will ask @Aslemammad to release ;)
